### PR TITLE
fix(core): Breadcrumbs added on forked context are now captured

### PR DESCRIPTION
**Pull request by antonis**
## :loudspeaker: Type of change

- [x] Bugfix
- [ ] New feature
- [ ] Enhancement
- [ ] Refactoring

## :scroll: Description
Uses `breadcrumb.origin` field to prevent exception capture context from being overwritten by native scope sync.

* [Adds Capture exception with breadcrumb in the sample](https://github.com/getsentry/sentry-react-native/pull/4124/commits/37553f296a99d532ff69a63c1953247c67ec7d1e)
* [Sets `react-native` as breadcrumb origin when not set](https://github.com/getsentry/sentry-react-native/pull/4124/commits/24a527be83850ad4a63b47920b025fc1f7fe7003)
* [Filters out breadcrumbs with react-native origin from the native layer](https://github.com/getsentry/sentry-react-native/pull/4124/commits/fabf0526a8b9d70a75c703b2d29accbc09e4fe93)
* [Merges event breadcrumbs with native context](https://github.com/getsentry/sentry-react-native/pull/4124/commits/d0b163342380980a878c340f8ad55df08e86d1a5)
* Refactors and tests RNSentryModuleImpl.fetchNativeDeviceContexts (reviewed separately in https://github.com/getsentry/sentry-react-native/pull/4253)

Depends on:
- https://github.com/getsentry/sentry-java/pull/3727
- https://github.com/getsentry/sentry-cocoa/pull/4358
- https://github.com/getsentry/sentry-docs/pull/11449

## :bulb: Motivation and Context

Fixes https://github.com/getsentry/sentry-react-native/issues/2146
The discussion in https://github.com/getsentry/sentry-capacitor/pull/629#discussion_r1617321563 adds some context on the need of a new field to be used internally to solve a technical problem.

## :green_heart: How did you test it?
Manual testing with the `Capture exception with breadcrumb` button, CI

|Before|After iOS|After Android|
|---|---|---|
|main|ios|android|

## :pencil: Checklist

- [x] I reviewed submitted code
- [x] I added tests to verify changes
- [x] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled
- [x] All tests passing
- [x] No breaking changes

### CODE REVIEW

**Comment by lucas-zimerman** [comment]
> Hey @lucas-zimerman, @krystofwoldrich 👋
> 
> > Fixes #2146
> > The discussion in [getsentry/sentry-capacitor#629 (comment)](https://github.com/getsentry/sentry-capacitor/pull/629#discussion_r1617321563) adds some context on the need of a new field to be used internally to solve a technical problem.
> 
> Since some time has passed since the creation of the issue and the discussion above I'd like to verify that the direction I've chosen with this implementation is valid:
> 
>     * [Set react-native as breadcrumb origin](https://github.com/getsentry/sentry-react-native/pull/4124/commits/24a527be83850ad4a63b47920b025fc1f7fe7003)
> 
>     * [Filters out breadcrumbs with `react-native` origin from the native layer](https://github.com/getsentry/sentry-react-native/pull/4124/commits/fabf0526a8b9d70a75c703b2d29accbc09e4fe93)
> 
>     * [Merges event breadcrumbs with native context](https://github.com/getsentry/sentry-react-native/pull/4124/commits/d0b163342380980a878c340f8ad55df08e86d1a5)
> 
> 
> Any feedback is welcome 🙏

The approach looks good indeed, thank you!
One thing to note is we should respect `maxBreadcrumbs` and limit the merged breadcrumbs to the defined value.

**Comment by krystofwoldrich** [comment]
Thank you @antonis for checking this.

> [Set react-native as breadcrumb origin](https://github.com/getsentry/sentry-react-native/pull/4124/commits/24a527be83850ad4a63b47920b025fc1f7fe7003)
[Filters out breadcrumbs with react-native origin from the native layer](https://github.com/getsentry/sentry-react-native/pull/4124/commits/fabf0526a8b9d70a75c703b2d29accbc09e4fe93)
[Merges event breadcrumbs with native context](https://github.com/getsentry/sentry-react-native/pull/4124/commits/d0b163342380980a878c340f8ad55df08e86d1a5)

Yes, all tree steps look good.

+ What @lucas-zimerman mentioned.

File: packages/core/ios/RNSentry.mm [hunk 1]
```diff
@@ -382,6 +382,17 @@ - (NSDictionary*) fetchNativeStackFramesBy: (NSArray<NSNumber*>*)instructionsAdd
 
     [serializedScope setValue:contexts forKey:@"contexts"];
     [serializedScope removeObjectForKey:@"context"];
+        
+    // Remove react-native breadcrumbs
+    NSMutableArray<NSDictionary<NSString *, id> *> *breadcrumbs = [serializedScope[@"breadcrumbs"] mutableCopy];
+    for (NSInteger i = breadcrumbs.count - 1; i >= 0; i--) {
```
Comment (krystofwoldrich) [inline packages/core/ios/RNSentry.mm:388]: This is unsafe as we are removing the items based on index while using the index to iterate over the array.

File: samples/react-native/src/Screens/ErrorsScreen.tsx [hunk 2]
```diff
@@ -88,6 +88,13 @@ const ErrorsScreen = (_props: Props) => {
             Sentry.captureException(error);
           }}
         />
+        <Button
+          title="Capture exception with breadcrumb"
+          onPress={() => {
+            Sentry.captureException(new Error('Captured exception with breadcrumb'),
+              context => context.addBreadcrumb({ message: 'error with breadcrumb' }));
```
Comment (krystofwoldrich) [inline samples/react-native/src/Screens/ErrorsScreen.tsx:95]: This is a good way to try the breadcrumbs merging. 👍

File: packages/core/ios/RNSentry.mm (hunk 1 above)
Comment (krystofwoldrich) [reply packages/core/ios/RNSentry.mm]: True, the backward loop should be ok. Thank you for changing it for the predicate.

File: packages/core/src/js/integrations/devicecontext.ts [hunk 3]
```diff
@@ -83,7 +84,8 @@ async function processEvent(event: Event): Promise<Event> {
     ? native['breadcrumbs'].map(breadcrumbFromObject)
     : undefined;
   if (nativeBreadcrumbs) {
-    event.breadcrumbs = nativeBreadcrumbs;
+    const maxBreadcrumbs = getClient()?.getOptions().maxBreadcrumbs ?? 100; // Default is 100.
+    event.breadcrumbs = nativeBreadcrumbs.concat(event.breadcrumbs || []).slice(0, maxBreadcrumbs);
```
Comment (krystofwoldrich) [inline packages/core/src/js/integrations/devicecontext.ts:88]: h: We have to ensure we keep the last `maxBreadcrumbs` (the last 100). Please also add a test that capture this edge case.

File: packages/core/android/src/main/java/io/sentry/react/RNSentryModuleImpl.java [hunk 4]
```diff
@@ -905,6 +907,17 @@ public void fetchNativeDeviceContexts(Promise promise) {
...
+    if (currentScope != null) {
+      // Remove react-native breadcrumbs
+      Iterator<Breadcrumb> breadcrumbsIterator = currentScope.getBreadcrumbs().iterator();
+      while (breadcrumbsIterator.hasNext()) {
+        Breadcrumb breadcrumb = breadcrumbsIterator.next();
+        if ("react-native".equals(breadcrumb.getOrigin())) {
+          breadcrumbsIterator.remove();
+        }
+      }
+    }
```
Comment (krystofwoldrich) [inline packages/core/android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:919]: m: I'm not sure how testable `fetchNativeDeviceContexts` is at the moment, but can we add tests for this method. If it requires refactor first lets do it in a follow up PR.

File: packages/core/src/js/integrations/devicecontext.ts (hunk 3 above)
Comment (antonis) [reply packages/core/src/js/integrations/devicecontext.ts]: Good point @krystofwoldrich 👍 
I've updated the logic and tests with c4246d789b690da34fb47056f38c63e25064a9ee

File: packages/core/src/js/integrations/devicecontext.ts [hunk 5]
```diff
@@ -83,7 +84,11 @@ async function processEvent(event: Event): Promise<Event> {
     ? native['breadcrumbs'].map(breadcrumbFromObject)
     : undefined;
   if (nativeBreadcrumbs) {
-    event.breadcrumbs = nativeBreadcrumbs;
+    const maxBreadcrumbs = getClient()?.getOptions().maxBreadcrumbs ?? 100; // Default is 100.
```
Comment (lucas-zimerman) [inline packages/core/src/js/integrations/devicecontext.ts:87]: @krystofwoldrich Do you think it's worth caching this value? It is quite unlikely that this option will change after the SDK initializes.

File: packages/core/src/js/integrations/devicecontext.ts [hunk 6]
```diff
@@ -83,7 +84,11 @@ async function processEvent(event: Event): Promise<Event> {
...
+    event.breadcrumbs = nativeBreadcrumbs
```
Comment (lucas-zimerman) [inline packages/core/src/js/integrations/devicecontext.ts:88]: thats a nice clean solution

File: packages/core/android/src/main/java/io/sentry/react/RNSentryModuleImpl.java (hunk 4 above)
Comment (antonis) [reply packages/core/android/src/main/java/io/sentry/react/RNSentryModuleImpl.java:912]: Good idea @krystofwoldrich 👍 
I did a small refactoring and added tests with https://github.com/getsentry/sentry-react-native/pull/4253

File: packages/core/src/js/integrations/devicecontext.ts (hunk 5 above)
Comment (krystofwoldrich) [reply packages/core/src/js/integrations/devicecontext.ts:87]: I think it's fine without, it's just reading value from object.

But what we should do it read the client from the function args. To avoid the lookup of the global client.

We currently use only the first arg of the function.

https://github.com/getsentry/sentry-javascript/blob/79818ffc304041238b948c20db2aa9a25458e798/packages/types/src/integration.ts#L53 

```js
processEvent?(event: Event, hint: EventHint, client: Client): Event | null | PromiseLike<Event | null>;
```

**Comment by krystofwoldrich** [comment]
@antonis Thank you for the updates. It looks great, I have one last small comment regarding @lucas-zimerman caching question. Then it's good to 🚀

**Review by krystofwoldrich — APPROVED** [review]
LG 🚀 Thank you!

**Comment by krystofwoldrich** [comment]
This PR doesn't trigger our dangerous file warning, but I thought anyway we might want to release this as a shot beta before latest release.

**Review by lucas-zimerman — APPROVED** [review]
Looks good too, yeah, it would be great if this PR goes first on a beta version.

**Comment by antonis** [comment]
>  I thought anyway we might want to release this as a shot beta before latest release.

This sounds like a good idea ➕
